### PR TITLE
[ADD] contract_price_revision: new module to update contract prices e…

### PR DESCRIPTION
**Pull request by cubells**
…asily

cc @Tecnativa

### CODE REVIEW

**Comment by rafaelbn** [comment]
@cubells please review travis :red_circle:

File: contract_price_revision/README.rst [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+.. |badge4| image:: https://img.shields.io/badge/weblate-Translate%20me-F47D42.png
+    :target: https://translation.odoo-community.org/projects/contract-11-0/contract-11-0-contract_price_revision
+    :alt: Translate me on Weblate
+.. |badge5| image:: https://img.shields.io/badge/runbot-Try%20me-875A7B.png
+    :target: https://runbot.odoo-community.org/runbot/110/11.0
+    :alt: Try me on Runbot
+
+|badge1| |badge2| |badge3| |badge4| |badge5| 
+
+This module allows create revisions of contract line prices.
```
Comment (pedrobaeza) [inline contract_price_revision/README.rst]: s/allows/allows to

File: contract_price_revision/README.rst [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+.. contents::
+   :local:
+
+Usage
+=====
+
+To use this module:
+
+#. Go to **Invoicing > Documents > Customer Contracts** if you are a billing
+   user or go to **Sales > Orders > Contracts** if you are a sales user.
```
Comment (pedrobaeza) [inline contract_price_revision/README.rst]: This part is not possible without the module `contract_sale`, so remove it. This also applies to "Supplier Contracts".

File: contract_price_revision/README.rst [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+#. Select as many contracts as you want to update line prices.
+#. Click on Action button and execute the wizard **Duplicate Contract Line**.
```
Comment (pedrobaeza) [inline contract_price_revision/README.rst]: This name is misleading. Call it "Create revision of contract lines".

File: contract_price_revision/README.rst [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+#. Enter date start from which new price will be valid. You can enter date
```
Comment (pedrobaeza) [inline contract_price_revision/README.rst]: s/new price/the new price

File: contract_price_revision/README.rst [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+   end and percent to increase old contract lines.
```
Comment (pedrobaeza) [inline contract_price_revision/README.rst]: The percentage (not percent) should be required. It's not interesting to create the exact same line.

File: contract_price_revision/README.rst [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+#. By clicking on Apply button, a new contract line will be created with
+   a price increased accordingly to the percent entered. Old contract lines
```
Comment (pedrobaeza) [inline contract_price_revision/README.rst:48]: s/percent/percentage

File: contract_price_revision/README.rst [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+   will have a date end from which pricelist will be out of date.
```
Comment (pedrobaeza) [inline contract_price_revision/README.rst]: `will have as ending date the day before the entered date.`

File: contract_price_revision/models/account_analytic_account.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+class AccountAnalyticAccount(models.Model):
+    _inherit = "account.analytic.account"
+
+    @api.model
+    def _prepare_invoice_line(self, line, invoice_id):
+        line_obj = self.env['account.invoice.line']
+        invoice = self.env['account.invoice'].browse(invoice_id)
+        # Line with automatic price are not taken into account
+        if (line.automatic_price or (
```
Comment (pedrobaeza) [inline contract_price_revision/models/account_analytic_account.py]: Remove `automatic_price` from this condition and directly don't duplicate that kind of lines.

File: contract_price_revision/models/account_analytic_account.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+                not line.previous_info_id and
+                invoice.date_invoice < line.date_end) or (
```
Comment (pedrobaeza) [inline contract_price_revision/models/account_analytic_account.py]: What happens if `line.date_end = False`. That's why other tests are failing.

File: contract_price_revision/models/account_analytic_invoice_line.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+class AccountAnalyticINvoiceLine(models.Model):
+    _inherit = "account.analytic.invoice.line"
+
+    date_start = fields.Date(
+        string='Start Date',
+    )
+    date_end = fields.Date(
+        string='End Date',
+    )
+    previous_info_id = fields.Many2one(
```
Comment (pedrobaeza) [inline contract_price_revision/models/account_analytic_invoice_line.py]: Weird name... Better to call it `previous_revision_id`

File: contract_price_revision/wizards/contract_duplicate_line.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+
+class ContractDuplicateLineWizard(models.TransientModel):
+    _name = 'contract.duplicate.line.wizard'
+
+    date_start = fields.Date(
+        required=True,
+    )
+    date_end = fields.Date()
+    variation_percent = fields.Float(
```
Comment (pedrobaeza) [inline contract_price_revision/wizards/contract_duplicate_line.py]: required

File: contract_price_revision/models/account_analytic_account.py (hunk 9 above)
Comment (pedrobaeza) [reply contract_price_revision/models/account_analytic_account.py]: As we are introducing the new feature of starting and ending dates. I think we should check them unconditionally if they have been put manually or through the revision wizard.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: contract_price_revision/models/account_analytic_account.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+
+class AccountAnalyticAccount(models.Model):
+    _inherit = "account.analytic.account"
+
+    @api.model
+    def _prepare_invoice_line(self, line, invoice_id):
+        line_obj = self.env['account.invoice.line']
+        invoice = self.env['account.invoice'].browse(invoice_id)
+        # Line with automatic price are not taken into account
+        if (line.automatic_price or not line.date_end or (
```
Comment (pedrobaeza) [inline contract_price_revision/models/account_analytic_account.py]: This condition is still too much complicated: only take into account dates. Forget `automatic_price` and `previous_revision_id`. Suggestion:
```python
if (line.date_start and invoice.date_invoice < line.date_start) or
        (line.date_end and invoice.date_invoice > line.date_end):
    return line_obj
return super()...
```

Comment (cubells) [reply contract_price_revision/models/account_analytic_account.py]: done @pedrobaeza !

**Review by pedrobaeza — APPROVED** [review]

File: contract_price_revision/models/account_analytic_account.py [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+from odoo import api, models
+
+
+class AccountAnalyticAccount(models.Model):
+    _inherit = "account.analytic.account"
+
+    @api.model
+    def _prepare_invoice_line(self, line, invoice_id):
+        line_obj = self.env['account.invoice.line']
+        invoice = self.env['account.invoice'].browse(invoice_id)
```
Comment (yajo) [inline contract_price_revision/models/account_analytic_account.py:12]: You can use `browse(invoice_id, prefetch=self._prefetch)` for performance.

File: contract_price_revision/wizards/create_revision_line_views.xml [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+                    <group>
+                        <field name="date_start"/>
+                        <field name="date_end"/>
+                        <field name="variation_percent"/>
+                    </group>
+                </group>
+                <footer>
+                    <button string="Apply" name="action_apply"
+                            type="object" class="oe_highlight"/>
+                    or
```
Comment (yajo) [inline contract_price_revision/wizards/create_revision_line_views.xml:26]: :fire:

**Review by yajo — APPROVED** [review]

**Review by JordiBForgeFlow — APPROVED** [review]
:+1: Functional and technical review.

**Pull request merged**